src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/AttributePresenceFilterNode.cs

```diff
@@ -0,0 +1,295 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Internal.JitInterface;
+using Internal.NativeFormat;
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILCompiler.DependencyAnalysis.ReadyToRun
+{
+    public class AttributePresenceFilterNode : HeaderTableNode
+    {
+        private EcmaModule _module;
+        private ObjectData computedData;
+
+        public override int ClassCode => 56456113;
+
+        public AttributePresenceFilterNode(EcmaModule module, TargetDetails target)
+            : base(target)
+        {
+            this._module = module;
+        }
+
+        public override void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix);
+            sb.Append("__ReadyToRunAttributePresenceFilter");
+        }
+
+        private struct CustomAttributeEntry
+        {
+            public string TypeNamespace;
+            public string TypeName;
+            public int Parent;
+        }
+
+        private List<CustomAttributeEntry> GetCustomAttributeEntries()
+        {
+            MetadataReader reader = _module.MetadataReader;
+            List<CustomAttributeEntry> customAttributeEntries = new List<CustomAttributeEntry>();
+            foreach (var handle in reader.CustomAttributes)
+            {
+                CustomAttribute customAttribute = reader.GetCustomAttribute(handle);
+                EntityHandle customAttributeConstructorHandle = customAttribute.Constructor;
+                MethodDesc customAttributeConstructorMethodDesc = this._module.GetMethod(customAttributeConstructorHandle);
```

Comment:
Grabbing the `MethodDesc` can lead to an exception if the custom attribute cannot be resolved - it will tear down the compilation at this point. It's not a problem for CoreLib, but it can be a problem if we go beyond CoreLib because customers do like to compile incomplete garbage code (based on our experience with .NET Native).

A way to avoid that would be if this keeps operating on raw metadata (decode MemberRefs, MethodDefs, etc.). It's a bit of extra work. @davidwrighton Do we need to address that?

Reply:
I'd like to see the metadata reading logic put together. I don't think its likely to be an issue, but we might as well, and its not as though its complicated risky logic to write.

Reply:
Fixed